### PR TITLE
HPCC-26402 Protect node cache from exceptions before node fully loaded

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@mckellyln Changes so far - haven't spotted any problems with them..

**Review by mckellyln — COMMENTED** [review]
Looks good, just had the one comment about not DBGLOG()ing the exception.

**Review by mckellyln — COMMENTED** [review]
Just have two comments about location of ownedNode and if legacyMode needs the same protection.

**Review by jakesmith — APPROVED** [review]
@ghalliday - looks good, please squash

**Review by mckellyln — APPROVED** [review]
Looks good.

File: system/jhtree/jhtree.cpp [hunk 1]
```diff
@@ -2470,45 +2493,62 @@ CJHTreeNode *CNodeCache::getNode(INodeLoader *keyIndex, unsigned iD, offset_t po
...
-        if (likely(alreadyExists))
-        {
-            cacheHits++;
-            if (ctx) ctx->noteStatistic(hitStatId[cacheType], 1);
-            (*hitMetric[cacheType])++;
-        }
-        else
+        //If an exception is thrown before the node is cleanly loaded we need to remove the partially constructed
+        //node from the cache otherwise it may never get loaded, and can prevent items being removed from the cache
+        //note: noteStatistic() can throw an exception if a worked has been aborted...
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply system/jhtree/jhtree.cpp]: I suspect it should be "If a query has been aborted..."

**Pull request merged**